sklearn/metrics/pairwise.py

```diff
@@ -256,6 +264,166 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+
+        dist(X, Y) = (X.shape[1] * 1 / ((dot(NX, NYT)))) *
+                    (dot((X * X), NYT) - 2 * (dot(X, Y.T)) +
+                    dot(NX, (Y.T * Y.T)))
+
+    where NX and NYT represent the logical-not of the missing masks of
+    X and Y.T, respectively.This formulation zero-weights feature coordinates
+    with missing value in either vector in the pair and up-weights the
+    remaining coordinates.
+    Formula derived by: Shreya Bhattarai <shreya.bhattarai@gmail.com>
+
+    Breakdown of euclidean distance calculation between a vector pair x,y:
+
+        weight = Total # of coordinates / # of non-missing coordinates
+        dist(x,y) = sqrt(weight * sq. distance from non-missing coordinates)
+
+    This of course implies that if all coordinates are missing in either
+    vector in the pair then NaN is returned for that pair.
+
+    Read more in the :ref:`User Guide <metrics>`.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples_1, n_features)
+
+    Y : {array-like, sparse matrix}, shape (n_samples_2, n_features)
+
+    squared : boolean, optional
+        Return squared Euclidean distances.
+
+    missing_values : "NaN" or integer, optional (default=”NaN”)
+        Representation of missing value
+
+    copy : boolean, optional
+        Make and use a deep copy of X and Y (if Y exists)
+
+    Returns
+    -------
+    distances : {array, sparse matrix}, shape (n_samples_1, n_samples_2)
+
+    Examples
+    --------
+    >>> from sklearn.metrics.pairwise import masked_euclidean_distances
+    >>> nan = float("NaN")
+    >>> X = [[0, 1], [1, nan]]
+    >>> # distance between rows of X
+    >>> masked_euclidean_distances(X, X)
+    array([[ 0.        ,  1.41421356],
+           [ 1.41421356,  0.        ]])
+
+    >>> # get distance to origin
+    >>> masked_euclidean_distances(X, [[0, 0]])
+    array([[ 1.        ],
+           [ 1.41421356]])
+
+    See also
+    --------
+    paired_distances : distances betweens pairs of elements of X and Y.
+    """
+    # Check and except sparse matrices
+    if issparse(X) or (Y is not None and issparse(Y)):
+        raise ValueError(
+            "Missing value support for sparse matrices not added yet")
+
+    # NOTE: force_all_finite=False allows not only NaN but also inf/-inf
+    X, Y = check_pairwise_arrays(X, Y,
+                                 force_all_finite=False, copy=copy)
+    if (np.any(np.isinf(X.data)) or
+            (Y is not None and np.any(np.isinf(Y.data)))):
+        raise ValueError(
+            "+/- Infinite values are not allowed.")
+
+    # Check if any rows have only missing value
+    if np.any(_get_mask(X, missing_values).sum(axis=1) == X.shape[1])\
+            or (Y is not None and np.any(_get_mask(Y, missing_values).sum(
+                axis=1) == Y.shape[1])):
+        raise ValueError("One or more rows only contain missing values.")
+    #
+    # if kill_missing:
```

Comment:
So my thinking for this was that a row with all NaN will only introduce "unnecessary" NaN values in the distance matrix. It does not technically matter for this specific situation (or for kneighbors) since all it does is return a row (or column) with all NaN values, but I was like why keep it when it can potentially introduce issues down the analysis-chain and does not contribute anything useful. So the current implementation requires users to get rid of samples that have nothing but NaN values before they pass the dataset to masked_kneighbors or masked_euclidean_distance(). However, if you prefer to remove this check for any reason please let me know and I will get rid of it.